cranelift/codegen/src/opts.rs

```diff
@@ -0,0 +1,309 @@
+
+use crate::egraph::Analysis;
+use crate::egraph::FuncEGraph;
+use crate::egraph::MemoryState;
+pub use crate::egraph::{Node, NodeCtx};
+use crate::ir::condcodes;
+pub use crate::ir::condcodes::{FloatCC, IntCC};
+pub use crate::ir::immediates::{Ieee32, Ieee64, Imm64, Offset32, Uimm32, Uimm64, Uimm8};
+pub use crate::ir::types::*;
+pub use crate::ir::{
+    dynamic_to_fixed, AtomicRmwOp, Block, Constant, DynamicStackSlot, FuncRef, GlobalValue, Heap,
+    Immediate, InstructionImms, JumpTable, MemFlags, Opcode, StackSlot, Table, TrapCode, Type,
+    Value,
+};
+use crate::isle_common_prelude_methods;
+use crate::machinst::isle::*;
+use crate::trace;
+pub use cranelift_egraph::{Id, NewOrExisting, NodeIter};
+use cranelift_entity::{EntityList, EntityRef};
+use smallvec::SmallVec;
+use std::marker::PhantomData;
+
+pub type IdArray = EntityList<Id>;
+#[allow(dead_code)]
+pub type Unit = ();
+pub type Range = (usize, usize);
+
+pub type ConstructorVec<T> = SmallVec<[T; 8]>;
+
+mod generated_code;
+use generated_code::ContextIter;
+
+struct IsleContext<'a, 'b> {
+    egraph: &'a mut FuncEGraph<'b>,
+}
+
+const REWRITE_LIMIT: usize = 5;
+
+pub fn optimize_eclass<'a>(id: Id, egraph: &mut FuncEGraph<'a>) -> Id {
+    trace!("running rules on eclass {}", id.index());
+    egraph.stats.rewrite_rule_invoked += 1;
+
+    if egraph.rewrite_depth > REWRITE_LIMIT {
+        return id;
+    }
+    egraph.rewrite_depth += 1;
+
+    // Find all possible rewrites and union them in, returning the
+    // union.
+    let mut ctx = IsleContext { egraph };
+    let optimized_ids = generated_code::constructor_simplify(&mut ctx, id);
+    let mut union_id = id;
+    if let Some(mut ids) = optimized_ids {
+        while let Some(new_id) = ids.next(&mut ctx) {
+            if ctx.egraph.subsume_ids.contains(&new_id) {
+                trace!(" -> eclass {} subsumes {}", new_id, id);
+                ctx.egraph.stats.node_subsume += 1;
+                // Merge in the unionfind so canonicalization still
+                // works, but take *only* the subsuming ID, and break
+                // now.
+                ctx.egraph.egraph.unionfind.union(union_id, new_id);
+                union_id = new_id;
+                break;
+            }
+            ctx.egraph.stats.node_union += 1;
+            let old_union_id = union_id;
+            union_id = ctx
+                .egraph
+                .egraph
+                .union(&ctx.egraph.node_ctx, union_id, new_id);
+            trace!(
+                " -> union eclass {} with {} to get {}",
+                new_id,
+                old_union_id,
+                union_id
+            );
+        }
+    }
+    ctx.egraph.subsume_ids.clear();
+    trace!(" -> optimize {} got {}", id, union_id);
+    ctx.egraph.rewrite_depth -= 1;
+    union_id
+}
+
+pub(crate) fn store_to_load<'a>(id: Id, egraph: &mut FuncEGraph<'a>) -> Id {
+    if let Some(load_key) = egraph.egraph.classes[id].get_node() {
```

Comment:
The optimization entrypoints are invoked for every new eclass node created (i.e., every time an enode is added to an eclass), so we only need to (and in fact should only, to avoid redundancy) examine just the latest enode in the eclass. I'll add a comment regarding this!

> Is it necessary for correctness to add the Load node to the egraph if the store gets forwarded to the load? Hypothetically, could the caller pass load_op, load_ty, etc to this function, and entirely skip constructing the Node::Load if it succeeds? I'm not asking you to change that in this PR, I just want to know if there are invariants that would break if someone did that.

I think that's technically possible, yeah, although there's one important case it would pessimize (in compile time, not code quality): having the `Load` node in the egraph as well lets other later `Load`s (with the same memory state) deduplicate with it. If we did not add the `Load` node to the egraph at all, then later `Load`s would eventually still end up with the forwarded stored value if the original did, because `store_to_load` is deterministic; but we would redo the work each time to determine that.